text/0046-metadata-for-offline-signers.md

```diff
@@ -0,0 +1,295 @@
+
+`blake3` transformation of concatenated child nodes (`blake3(left + right)`) as merge procedure;
+
+### Complete Binary Merkle Tree construction protocol
+
+1. Leaves are numbered in ascending order. Leaf index is associated with corresponding chunk.
+2. Merge is performed using the leaf with highest index as right and leaf with second to highest index as left children; result is pushed to the end of nodes queue and leaves are discarded.
+3. Step (2) is repeated until no leaves or just one leaf remains; in latter case, the last leaf is pushed to the front of the nodes queue.
+4. Right node and then left node is popped from the front of the nodes queue and merged; the result is sent to the end of the queue.
+5. Step (4) is repeated until only one node remains; this is tree root.
+
+
+```
+queue = empty_queue
+
+while (leaves.length>1) {
+  right = leaves.pop_last
+  left = leaves.pop_last
+  queue.push_back(merge(left, right))
+}
+
+if leaves.length == 1 {
+  queue.push_front(leaves.last)
+}
+
+while queue.len() > 1 {
+  right = queue.pop_front
+  left = queue.pop_front
+  queue.push_back(merge(left, right))
+}
+
+return queue.pop
+```
+
+```
+Resulting tree for metadata consisting of 5 nodes (numbered from 0 to 4):
+
+       root
+     /     \
+    *       *
+   / \     / \
+  *   0   1   2
+ / \
+3   4
+```
+
+### Digest
+
+1. Blake3 hash is computed for each chunk of modular short metadata registry.
+3. Complete Binary Merkle Tree is constructed as described above.
+4. Root hash of this tree (left) is merged with metadata descriptor blake3 hash (right); this is metadata digest.
+
+Version number and corresponding resulting metadata digest MUST be included into Signed Extensions as specified in Chain Verification section below.
+
+### Chain verification
+
+The root of metadata computed by cold device MAY be included into Signed Extensions; if it is included, the transaction will pass as valid iff hash of metadata as seen by cold storage device is identical to consensus hash of metadata, ensuring fair signing protocol.
+
+The Signed Extension representing metadata digest is a single byte representing both digest vaule inclusion and shortening protocol version; this MUST be included in Signed Extensions set. Depending on its value, a digest value is included as `additionalSigned` to signature computation according to following specification:
+
+| signed extension value | digest value   | comment                            |
+|------------------------|----------------|------------------------------------|
+| `0x00`                 |                | digest is not included             |
+| `0x01`                 | 32-byte digest | this represents protocol version 1 |
+| `0x02` - `0xFF`        | *reserved*     | reserved for future use            |
+
+## Drawbacks
+
+### Increased transaction size
+
+A 1-byte increase in transaction size due to signed extension value. Digest is not included in transferred transaction, only in signing process.
```

Comment:
we do already have a version byte in extrinsic format. maybe we can just use it instead of a new byte?